src/dataset-builder/ConceptSearch.test.ts

```diff
@@ -0,0 +1,152 @@
+    render(
+      h(ConceptSearch, {
+        actionText,
+        initialSearch,
+        initialCart,
+        onCancel,
+        onCommit,
+        onOpenHierarchy,
+        datasetId,
+        domainOption,
+      })
+    );
+  };
+
+  const displayedConcepts = [dummyGetConceptForId(102), dummyGetConceptForId(103)];
+  const mockSearch = jest.fn();
+
+  const mockDataRepoContract: DataRepoContract = {
+    dataset: (_datasetId) =>
+      ({
+        searchConcepts: mockSearch,
+      } as Partial<DataRepoContract['dataset']>),
+  } as Partial<DataRepoContract> as DataRepoContract;
+  asMockedFn(DataRepo).mockImplementation(() => mockDataRepoContract as DataRepoContract);
+  mockSearch.mockResolvedValue({
+    result: displayedConcepts,
+  });
+
+  it('renders the concept search with an initial search result', async () => {
+    // Arrange
+    renderSearch();
+    // Assert
+    expect(await screen.findByText(displayedConcepts[0].name)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[0].id)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[1].name)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[1].id)).toBeTruthy();
+    // Action text not visible until a row is selected.
+    expect(screen.queryByText(actionText)).toBeFalsy();
+  });
+
+  it('searches based on user input', async () => {
+    // Arrange
+    renderSearch();
+    const searchText = 'search text';
+    // Act
+    const user = userEvent.setup();
+    await user.type(screen.getByPlaceholderText('Search'), searchText);
+    // Assert - the search is called with the search text.
+    expect(mockSearch).toHaveBeenCalledWith(domainOption.root, searchText);
+  });
+
+  it('filters based on search text', async () => {
+    // Arrange
+    renderSearch(displayedConcepts[0].name);
+    // Act
+    // Assert
+    expect(screen.findByText(displayedConcepts[0].name)).toBeTruthy();
+    // The second concept should not be displayed.
+    expect(() => screen.getByText(displayedConcepts[1].name)).toThrow();
+  });
+
+  it('calls open hierarchy on click', async () => {
+    // Arrange
+    const concept = displayedConcepts[0];
+    renderSearch();
+    await screen.findByText(concept.name);
+    // Act
+    const user = userEvent.setup();
+    await user.click(screen.getByLabelText(`open hierarchy ${concept.id}`));
+    // Assert
+    expect(onOpenHierarchy).toHaveBeenCalledWith(
+      { id: concept.id, category: domainOption.category, root: concept },
+      [],
+      ''
+    );
+  });
+
+  it('supports add to cart', async () => {
+    // Arrange
+    renderSearch();
+    const concept = displayedConcepts[0];
+    await screen.findByText(concept.name);
+    // Act
+    const user = userEvent.setup();
+    await user.click(screen.getByLabelText(`add ${concept.id}`));
+    // Assert
+    expect(screen.queryByText(actionText)).toBeTruthy();
+    expect(screen.queryByText('1 concept', { exact: false })).toBeTruthy();
+  });
```

Comment:
would it be possible to add a test that shows that the cart persists across the search and hierarchy pages? Like when you go back to the search page your concept should still be selected

Reply:
I thought about that; there are a few tests possible:
- assert that the current cart is passed to the new page on transition; it could be added to calls open hierarchy on click`
- assert that when a cart is passed in, it's used to populate the initial page state

It's not hard to test these cases too. I do wonder about how much testing is really useful here..

Reply:
I added a few more tests